File: haskell/app/Game.hs
module Game where

import GHC.Arr
import Data.Array.IO
import Data.List
import Data.Maybe

import Canvas
import Input (Input)

import Game.Draw qualified as Draw  
import Game.Field
import Game.Input
import Game.Piece hiding (dims)
import Game.Piece qualified as Piece (dims) 
import Game.Timer

data Data = Data
  { state              :: State
  , field              :: Field
  , piece              :: Piece
  , gravityTimer       :: Timer
  , clearRowFlashTimer :: Timer
  } deriving Show

data State
  = SpawnPiece
  | Spawned { pos :: (Int, Int) }
  | BlinkDisappearingRows { doShow :: Bool, blinks :: Int }
  | ClearDisappearingRows
  | FallAfterClear
  | ChangePiece
  deriving Show

gravityTick :: Double
gravityTick = 0.5

mkData :: Data
mkData = Data
  { state = SpawnPiece
  , field = mkField
  , piece = mkPiece
  , gravityTimer = mkTimer gravityTick
  , clearRowFlashTimer = mkTimer (gravityTick / 2.0)
  }

spawnPiece :: Data -> Data
spawnPiece d = 
  let pos = (0, 5)
  in d { state = Spawned { pos = pos }, field = spawn (piece d) pos (field d) }

spawned :: Maybe Turn -> Maybe Move -> Bool -> (Int, Int) -> Data -> Data
spawned turn mov hasTicked pos d = 
  let
    (newPiece, newField, newPos) =
      let
        (piece', field', pos') = fromMaybe (piece d, field d, pos) (turn    >>= tryTurnPiece pos  (piece d) (field d))
        (field'', pos'')       = fromMaybe (defaultField, pos')    (moveDir >>= tryMovePiece pos' piece'    field'   )
          where
            (defaultField, moveDir) =
              if hasTicked
              then (amap fallingToFrozen field', Just MoveDown)
              else (field'                     , mov          )
            fallingToFrozen Falling = Frozen
            fallingToFrozen c       = c 
      in
        (piece', field'', pos'') 
    rows        = fst $ rowsCols newField
    hasFullRows = isJust . find (all (==Frozen)) $ rows
    newField'   = newField // [(Pos (y, x), Disappearing) | y <- frozenRowYs, x <- frozenRowXs]
      where
        frozenRowYs = map fst . filter (all (==Frozen) . snd) $ zip [0..] rows
        frozenRowXs = [0..snd (dims newField) - 1]
  in d
    { state =
      if hasFullRows                     then BlinkDisappearingRows { doShow = False, blinks = 4 }
      else if hasTicked && newPos == pos then ChangePiece
      else                                    Spawned { pos = newPos }
    , piece = newPiece
    , field = newField'
    }

blinkDisappearingRows :: Bool -> Int -> Double -> Data -> Data
blinkDisappearingRows doShow blinks dt d =
  if blinks > 0
  then
    let
      (clearRowFlashTimer', doFlash) = tick dt $ clearRowFlashTimer d
    in d
      { clearRowFlashTimer = clearRowFlashTimer'
      , state =
          if doFlash
          then BlinkDisappearingRows { doShow = not doShow, blinks = blinks - 1 }
          else BlinkDisappearingRows { doShow = doShow    , blinks = blinks     }
      }
  else d { state = ClearDisappearingRows }

clearDisappearingRows :: Data -> Data 
clearDisappearingRows d = 
  d { state = FallAfterClear, field = amap clear $ field d }
  where
    clear Frozen       = Falling
    clear Disappearing = Empty
    clear c            = c

fallAfterClear :: Data -> Data
fallAfterClear d =
  case tryFall $ field d of
    Nothing -> d { state = ChangePiece, field = amap fallingToFrozen (field d) }
      where
        fallingToFrozen Falling = Frozen
        fallingToFrozen c       = c
    Just field' -> d { field = field' }

update :: Data -> Canvas -> Input -> Double -> IO (Data, Canvas)
update d canvas input dt = do
  let
    mov                        = mkMove input
    turn                       = mkTurn input
    (gravityTimer', hasTicked) = tick dt (gravityTimer d)
    d' = case state d of
      SpawnPiece                               -> spawnPiece d
      Spawned { pos }                          -> spawned turn mov hasTicked pos d
      BlinkDisappearingRows { doShow, blinks } -> blinkDisappearingRows doShow blinks dt d
      ClearDisappearingRows                    -> clearDisappearingRows d
      FallAfterClear                           -> if hasTicked then fallAfterClear d else d
      ChangePiece                              -> d { state = SpawnPiece, piece = next $ piece d }

  -- draw
  Draw.clear canvas

  (_, (ch, cw)) <- getBounds canvas
  let
    cellPx = 30
    drawDisappearing =
      case state d' of
        BlinkDisappearingRows { doShow } -> doShow
        _                                -> False
    (_, Pos (fh, fw)) = bounds $ field d'
    (y0, x0) = (ch `div` 2 - (fh * cellPx) `div` 2, cw `div` 2 - (fw * cellPx) `div` 2)

  Draw.field (field d') (y0, x0) cellPx drawDisappearing canvas
  Draw.field (blueprint $ next $ piece d') (y0, x0 + fw * cellPx + cellPx) cellPx False canvas
  pure (d' { gravityTimer = gravityTimer' }, canvas)

spawn :: Piece -> (Int, Int) -> Field -> Field
spawn piece pos = copyIf (==Falling) (blueprint piece) pos (Piece.dims piece)

tryFall :: Field -> Maybe Field
tryFall field = do
  y <- find isFloating . drop 1 . reverse $ [0..height - 1]
  let ys = [0..y]
  Just $ field
    // [(Pos (y    , x), Empty             ) | y <- ys, x <- xs]
    // [(Pos (y + 1, x), field ! Pos (y, x)) | y <- ys, x <- xs]
  where
    (height, width) = dims field

    xs = [0..width - 1]

    isFloating y = any (\x -> field ! Pos (y    , x) == Falling) xs
                && all (\x -> field ! Pos (y + 1, x) == Empty  ) xs

tryMovePiece :: (Int, Int) -> Piece -> Field -> Move -> Maybe (Field, (Int, Int))
tryMovePiece pos piece field m = do
  pos' <-
    let
      (pieceHeight, pieceWidth) = Piece.dims piece
      (y', x') = pos
      (y, x) = case m of
        MoveLeft  -> (y'    , x' - 1)
        MoveRight -> (y'    , x' + 1)
        MoveDown  -> (y' + 1, x'    )
    in
      if all (inRange (bounds field) . Pos) [(y, x), (y + pieceHeight - 1, x + pieceWidth - 1)]
      then Just (y, x)
      else Nothing
  if hasCollided piece pos' field
  then Nothing
  else Just (movePiece piece pos' field, pos')

tryTurnPiece :: (Int, Int) -> Piece -> Field -> Turn -> Maybe (Piece, Field, (Int, Int))
tryTurnPiece pos piece field t =
  let
    turned = turn t piece
    (pieceHeight, pieceWidth) = Piece.dims turned
    pos' =
      let
        (y, x) = pos
        (height, width) = dims field
      in
        ( if y + pieceHeight > height then height - pieceHeight else y
        , if x + pieceWidth  > width  then width  - pieceWidth  else x
        )
  in
    if hasCollided turned pos' field
    then Nothing
    else Just (turned, movePiece turned pos' field, pos')

movePiece :: Piece -> (Int, Int) -> Field -> Field
movePiece piece pos = copyIf (==Falling) (blueprint piece) pos (Piece.dims piece) . amap fallingToEmpty
  where
    fallingToEmpty Falling = Empty
    fallingToEmpty c       = c

hasCollided :: Piece -> (Int, Int) -> Field -> Bool
hasCollided piece pos field =
  isJust . find collides $
    [ (blueprint piece ! Pos (bpY, bpX) , field ! Pos (y, x))
    | (bpY, y) <- zip [0..] [y0..y1]
    , (bpX, x) <- zip [0..] [x0..x1]
    ]
  where
    (y0, x0) = pos
    (y1, x1) = (y0 + pieceHeight - 1, x0 + pieceWidth - 1)
      where
        (pieceHeight, pieceWidth) = Piece.dims piece

    collides (Falling, Falling) = False
    collides (Falling, Empty  ) = False
    collides (Falling, _      ) = True
    collides (_      , _      ) = False

File: rust/src/game.rs
mod field;
mod draw;
mod grid;
mod input;
mod piece;
mod timer;

use std::any::Any;

use field::{Field, Cell};
use draw::Canvas;
use grid::Grid;
use input::{Move, Turn};
use piece::Piece;
use timer::Timer;

use crate::lume::{Input, RawCanvas};

pub struct Data {
    state: State,
    field: Field,
    piece: Piece,
    gravity_timer: Timer,
    clear_row_flash_timer: Timer,
}

#[derive(Copy, Clone, PartialEq, Debug)]
enum State {
    SpawnPiece,
    Spawned {
        pos: [usize; 2],
    },
    BlinkDisappearingRows {
        show: bool,
        blinks: i32,
    },
    ClearDisappearingRows,
    FallAfterClear,
    ChangePiece,
}

impl Default for Data {
    fn default() -> Self {
        const GRAVITY_TICK: f64 = 0.5;

        Self {
            state: State::SpawnPiece,
            field: Grid::filled(Cell::Empty),
            piece: Piece::new(),
            gravity_timer: Timer::new(GRAVITY_TICK),
            clear_row_flash_timer: Timer::new(GRAVITY_TICK / 2.0),
        }
    }
}

pub fn update(data: &mut dyn Any, raw_canvas: &mut dyn RawCanvas, input: &Input, dt: f64) {
    let data = data.downcast_mut::<Data>().unwrap();
    let mut canvas = Canvas::from_raw(raw_canvas);

    inner(data, &mut canvas, input, dt);
}

fn inner(data: &mut Data, canvas: &mut Canvas, input: &Input, dt: f64) {
    let tick = data.gravity_timer.tick(dt);
    let mov = if tick {
        Some(Move::Down)
    } else {
        Option::<Move>::from(input)
    };
    let turn = Option::<Turn>::from(input);

    let mut show_disappearing = true;

    data.state = match data.state {
        State::SpawnPiece => {
            let pos = [0, 5];
            spawn(&mut data.field, data.piece, pos);

            State::Spawned { pos }
        }
        State::Spawned { pos } => {
            let mut new_pos = turn.and_then(|t| try_turn_piece(data, t, pos)).unwrap_or(pos);
            new_pos = mov.and_then(|m| try_move_piece(data, m, new_pos)).unwrap_or_else(|| {
                if tick {
                    data.field.iter_mut()
                        .filter(|c| **c == Cell::Falling)
                        .for_each(|c| *c = Cell::Frozen);
                }
                new_pos
            });

            let mut has_full_rows = false;
            data.field.rows_mut()
                .filter(|row| row.iter().all(|c| *c == Cell::Frozen))
                .for_each(|row| {
                    for cell in row {
                        *cell = Cell::Disappearing;
                    }

                    has_full_rows = true;
                });

            if has_full_rows {
                State::BlinkDisappearingRows { show: false, blinks: 4 }
            } else if tick && new_pos == pos {
                State::ChangePiece
            } else {
                State::Spawned { pos: new_pos }
            }
        }
        same @ State::BlinkDisappearingRows { show, blinks } => {
            if blinks > 0 {
                show_disappearing = show;

                if data.clear_row_flash_timer.tick(dt) {
                    State::BlinkDisappearingRows { show: !show, blinks: blinks - 1 }
                } else {
                    same
                }
            } else {
                show_disappearing = false;

                State::ClearDisappearingRows
            }  
        }
        State::ClearDisappearingRows => {
            for cell in data.field.iter_mut() {
                *cell = match *cell {
                    Cell::Frozen => Cell::Falling,
                    Cell::Disappearing => Cell::Empty,
                    c => c,
                }
            }

            State::FallAfterClear
        }
        same @ State::FallAfterClear => {
            if tick && !try_fall(&mut data.field) {
                data.field.iter_mut()
                    .filter(|c| **c == Cell::Falling)
                    .for_each(|c| *c = Cell::Frozen);

                State::ChangePiece
            } else {
                same
            }
        }
        State::ChangePiece => {
            data.piece = data.piece.next();

            State::SpawnPiece
        }
    };

    // draw
    {
        const CELL_PX: usize = 30;

        draw::clear(canvas);

        let bucket_width_px = data.field.width() * CELL_PX;
        let bucket_height_px = data.field.height() * CELL_PX;

        let x0 = canvas.width() / 2 - bucket_width_px / 2;
        let y0 = canvas.height() / 2 - bucket_height_px / 2;
        draw::grid(canvas, &data.field, CELL_PX, [y0, x0], show_disappearing);

        draw::grid(canvas, &data.piece.next().blueprint(), CELL_PX, [y0, x0 + bucket_width_px + CELL_PX], false);

        let mouse = &input.mouse;
        draw::cell(canvas, CELL_PX, [mouse.y, mouse.x], draw::color::GREEN);
    }
}

fn spawn(cells: &mut Field, piece: Piece, pos: [usize; 2]) {
    Grid::copy_if(cells, &piece.blueprint(), pos, piece.dims(), |c| *c == Cell::Falling);
}

fn try_fall(cells: &mut Field) -> bool {
    let mut fell = false;

    let width = cells.width();
    let height = cells.height();
    for y in (0..height).rev().skip(1) {
        let dst_is_free = (0..width).map(|x| cells[[y + 1, x]]).all(|cell| cell == Cell::Empty);
        let src_has_falling = (0..width).map(|x| cells[[y, x]]).any(|cell| cell == Cell::Falling);

        if dst_is_free && src_has_falling {
            for x in 0..width {
                cells[[y + 1, x]] = cells[[y, x]];
                cells[[y, x]] = Cell::Empty;
            }
            fell = true;
        }
    }

    fell
}

fn try_move_piece(data: &mut Data, mov: Move, pos: [usize; 2]) -> Option<[usize; 2]> {
    let new_pos = {
        let [height, width] = [data.field.height(), data.field.width()];
        let [piece_height, piece_width] = data.piece.dims();

        let [y, x] = pos;
        let [y, x] = match mov {
            Move::Left => [y, x.wrapping_sub(1)],
            Move::Right => [y, x + 1],
            Move::Down => [y + 1, x],
        };
        if y >= height || x >= width || y + piece_height > height || x.wrapping_add(piece_width) > width {
            return None
        }

        [y, x]
    };

    if has_collided(&data.piece, &data.field, new_pos) {
        return None
    }

    move_piece(&mut data.field, &data.piece, new_pos);
    Some(new_pos)
}

fn try_turn_piece(data: &mut Data, turn: Turn, pos: [usize; 2]) -> Option<[usize; 2]> {
    let mut turned = data.piece;
    turned.turn(turn);

    let new_pos = {
        let [height, width] = [data.field.height(), data.field.width()];
        let [piece_height, piece_width] = turned.dims();

        let [y, x] = pos;
        [
            if y + piece_height > height { height - piece_height } else { y },
            if x + piece_width > width { width - piece_width } else { x },
        ]
    };

    if has_collided(&data.piece, &data.field, new_pos) {
        return None
    }

    data.piece = turned;
    move_piece(&mut data.field, &data.piece, new_pos);
    Some(new_pos)
}

fn has_collided(piece: &Piece, field: &Field, [y0, x0]: [usize; 2]) -> bool {
    let [piece_height, piece_width] = piece.dims();
    let blueprint = piece.blueprint();

    let [y1, x1] = [y0 + piece_height, x0 + piece_width];

    for (bp_y, y) in (y0..y1).enumerate() {
        for (bp_x, x) in (x0..x1).enumerate() {
            let src = blueprint[[bp_y, bp_x]];
            let dst = field[[y, x]];

            if src == Cell::Falling && !(dst == Cell::Falling || dst == Cell::Empty) {
                return true
            }
        }
    }

    false
}

fn move_piece(field: &mut Field, piece: &Piece, pos: [usize; 2]) {
    field.iter_mut()
        .filter(|c| **c == Cell::Falling)
        .for_each(|c| *c = Cell::Empty);

    field.copy_if(&piece.blueprint(), pos, piece.dims(), |c| *c == Cell::Falling);
}